hazelcast/src/test/java/com/hazelcast/jet/core/JobUploadClientFailureTest.java

```diff
@@ -0,0 +1,296 @@
+    }
+
+    @Test
+    public void testTooLongFileName() {
+        HazelcastInstance client = createResourceUploadEnabledMemberAndClient();
+        JetService jetService = client.getJet();
+        JetClientInstanceImpl spyJetService = (JetClientInstanceImpl) Mockito.spy(jetService);
+
+        Path jarPath = getJarPath();
+
+        // Create a repeated filename
+        char[] charArray = new char[300];
+        Arrays.fill(charArray, 'a');
+        String longFileName = new String(charArray);
+        // JetClientInstanceImpl.getFileNameWithoutExtension return the long file name
+        when(spyJetService.getFileNameWithoutExtension(jarPath)).thenReturn(longFileName);
+
+        SubmitJobParameters submitJobParameters = new SubmitJobParameters()
+                .setJarPath(jarPath);
+
+        assertThatThrownBy(() -> spyJetService.submitJobFromJar(submitJobParameters))
+                .isInstanceOf(JetException.class)
+                .hasCauseInstanceOf(FileSystemException.class);
+    }
+
+    @Test
+    public void test_jarUpload_whenResourceUploadIsNotEnabled() {
+        createHazelcastInstance();
+        HazelcastInstance client = createHazelcastClient();
+        JetService jetService = client.getJet();
+
+        SubmitJobParameters submitJobParameters = new SubmitJobParameters()
+                .setJarPath(getJarPath());
+
+        assertThrows(JetException.class, () ->
+                jetService.submitJobFromJar(submitJobParameters)
+        );
+
+        assertEqualsEventually(() -> jetService.getJobs().size(), 0);
+    }
+
+    @Test
+    public void test_jarUpload_withWrongMainClassname() {
+        HazelcastInstance client = createResourceUploadEnabledMemberAndClient();
+        JetService jetService = client.getJet();
+
+        SubmitJobParameters submitJobParameters = new SubmitJobParameters()
+                .setJarPath(getJarPath())
+                .setMainClass("org.example.Main1");
+
+        assertThrows(JetException.class, () -> jetService.submitJobFromJar(submitJobParameters));
+    }
+
+
+    @Test
+    public void test_jar_isDeleted() throws IOException {
+        // Delete left over files before the test
+        String newSimpleJob = "newsimplejob.jar";
+        deleteLeftOverFilesIfAny(newSimpleJob);
+
+        HazelcastInstance client = createResourceUploadEnabledMemberAndClient();
+        JetService jetService = client.getJet();
+
+        // Copy as new jar to make it unique
+        Path newPath = copyJar(newSimpleJob);
+
+        SubmitJobParameters submitJobParameters = new SubmitJobParameters()
+                .setJarPath(newPath)
+                .setMainClass("org.example.Main1");
+
+        assertThrows(JetException.class, () -> jetService.submitJobFromJar(submitJobParameters));
+
+        boolean fileDoesNotExist = fileDoesNotExist(newSimpleJob);
+        assertThat(fileDoesNotExist)
+                .isTrue();
+
+        // Delete local new jar
+        Files.delete(newPath);
```

Comment:
If it is not cleaned up we want the test to fail.

Reply:
yes, of course we want the test to fail if it's not cleaned up - and it will fail because of previous assertion not because of this line. This line serves as clean up after the test. What I'm saying is that if the test fail then this line is not executed and file on `newPath` will not be deleted.